### PR TITLE
Fix whitespace parsing between expressions

### LINKED ISSUE #1119: Incorrect parsing (as function call) when something in braces is in line next to expression
### Instructions

Please fill out the template below to the best of your ability and include a label indicating which tool/service you were working with when you encountered the problem.

### Problem

This code

```
pre {
    (royaltyReceiver == nil) == (royaltyPercent == nil)
    (royaltyPercent == nil ) || (royaltyPercent! <= self.maxRoyaltyPercent)
}
```

raises error:

```
error: cannot call type: `Bool`
E             --> f8d6e0586b0a20c7.XXX:57:13
E              |
E           57 |             (royaltyReceiver == nil) == (royaltyPercent == nil)
E              |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

While other code with same meaning works without interpreter errors:

```
 pre {
    payment.balance == self.priceFor(tokenID: tokenID)
    self.isNFTStored(tokenID: tokenID)
}
```

### Steps to Reproduce

Version of emulator used:

```
flow version  
Version: v0.25.0
Commit: 9afd7938ebf603611d7b159278a259c850e0adda
```

### Acceptance Criteria

To me it looks like parsing ambiguity bug, I think code should be accepted.

### PR DESCRIPTION
**Pull request by turbolent**
Closes #1119 

## Description

Fix the parsing of whitespace/newlines between expressions. Even though we had logic in place to prevent parsing when a newline is encountered, it did not handle all cases and was also unnecessarily complicated.
The issue was that the space token was always consumed, but should only be conditionally skipped.

While at it:
- Improve the parsing of array and dictionary expressions
- Improve the AST explorer:
  - Update the dependencies
  - Switch from using a WASM binary of the parser to a Go HTTP API, which allows easier iteration: after an update of the parser, only the API has to be restarted, which is fast and can be done automatically using a file change watcher

This is pretty much a bugfix, and existing code should keep parsing as-is. I will verify this by validating the Mainnet contracts.
However, given that parsing behaviour changed (for the better), this change must theoretically still be seen as a breaking change, as existing code could depend on the old parsing behaviour.
We might need to/want to make this a Stable Cadence change.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by robert-e-davidson3 — APPROVED** [review]

File: tools/astexplorer/src/index.tsx [hunk 1]
```diff
@@ -75,20 +79,53 @@ document.addEventListener('DOMContentLoaded', async () => {
     }
   );
 
-  const parser = await CadenceParser.create("cadence-parser.wasm")
+  const root = createRoot(astElement)
+  const render = root.render.bind(root)
```
Comment (dreamsmasher) [inline tools/astexplorer/src/index.tsx:83]: you can use `root::render` with @babel/plugin-proposal-function-bind. but we also don't use babel for the ast explorer 😅

**Review by dreamsmasher — APPROVED** [review]

**Review by SupunS — APPROVED** [review]
LGTM!

**Comment by turbolent** [comment]
Follow-up: I analyzed all Mainnet contracts and noticed some existing contracts broke. This is fixed with #2136. With the fix, all contracts that currently parse on Mainnet still do, and parsing results are the same for them after this PR. 

TLDR: Theoretically a breaking change, but no impact on Mainnet contracts.
